### DISCUSSION TITLE
How to search for android theme properties

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13122] [Q&A] [answered]
I tried to change 3 dots option menu color and search a lot to find a solution that told me to use `ThemeOverlay.MaterialComponents.Toolbar.Primary` instead `ThemeOverlay.MaterialComponents.ActionBar.Surface`. Till now I don't know how find the right parent class in general to customize my Actionbar or Toolbar. For example the property `<item name="android:fontFamily">@font/...</item>` change the font-familey in both classes actionbar and toolbar. But it changes also the font of the options menu-items. How can I find the information in material code reference which classes to use if I want to distinguish the font-familey for actionbar-title and option menu-items?

**Comment by m1ga** [comment, marked as answer]
you've lnked to the correct articles before. 
```xml
<item name="actionBarTheme">@style/Theme.ActionBarTheme</item>
```
with
```xml
<style name="Theme.ActionBarTheme" parent="@style/Widget.AppCompat.Light.ActionBar">
	<item name="colorControlNormal">#f0f</item>
</style>
```
makes it look like this:
![Screenshot_20211015-151650](https://user-images.githubusercontent.com/4334997/137493057-66e676fc-0795-495d-b596-e9d99e5a9824.png)

Since those parts are mainly native Android themes you have to keep searching there `android theme actionbar fontfamily` or so.

For fonts i would try this in your theme
```
<item name="actionBarStyle">@style/CustomTheme.ActionBarStyle</item>
```

with
```xml
<style name="CustomTheme.ActionBarStyle" parent="@style/Widget.AppCompat.Light.ActionBar.Solid.Inverse">
	<!-- actionbar -->
	<item name="elevation">0dp</item>
	<item name="android:actionMenuTextColor">#f00</item>
	<item name="android:titleTextStyle">@style/CustomTheme.ActionBar.TitleTextStyle</item>
	<item name="titleTextStyle">@style/CustomTheme.ActionBar.TitleTextStyle</item>
</style>
```
and
```xml
<style name="CustomTheme.ActionBar.TitleTextStyle" parent="@android:style/TextAppearance.Holo.Widget.ActionBar.Title">
	<!-- actionbar font -->
	<item name="android:textColor">#00f</item>
</style>
```

Found it somewhere on Stackoverflow too. I've never made a native App so I'm not familiar with the themes either but there is a lot of code out there (some is very old and might not work).
If you want to see where the fonts have to be: https://github.com/appcelerator/titanium_mobile/discussions/13060 a small tutorial for the Tabbar font

**Reply by m1ga** [reply]
about the `parent` parts: most of the time it can have different parents. With different parents it will get a different base-look. It's trail & error sometimes

**Reply by beppo-ivel** [reply]
> parent="@style/Widget.AppCompat.Light.ActionBar.Solid.Inverse"
I thought *.AppCombat.* should be used only for Titanium versions older than 9.3.0?

> I've never made a native App so I'm not familiar with the themes
How do u customize your actionbar/toolbar or textfields? How do u create the components like actionbar or textfields that are not customizable via *.tss by default?

**Reply by m1ga** [reply]
I'm using `Widget.AppCompat` not `Theme.AppCompat` which is deprecated. But it might work with other parents too. 
I style the actionbar with a theme like shown above (I google the elements) and most of the time it is just the color (https://titaniumsdk.com/guide/Titanium_SDK/Titanium_SDK_How-tos/User_Interface_Deep_Dives/Android_UI_Components_and_Conventions/Android_Themes.html#android-material-themes).
Textfields: never had a problem with just using TSS properties :smile: Even the Tabbar tutorial above was made for different user. If my Tabbar is not looking like a native Tabbar I create a custom view with buttons + a ScrollableView so I have full control over the look and feel (e.g. add a dot for an active tab, add badgest,...). Setup is quicker than searching on alternatives on Google :wink: Performance: no difference and it works on both platforms right away

**Reply by m1ga** [reply]
Custom "Actionbar" with different size, background image, logo, hamburger menu, settings icon and a custom bottom navigation (did that before the TabGroup could be a bottom navigation):
![Screenshot_20211016-152046](https://user-images.githubusercontent.com/4334997/137589204-c0432153-c89d-4529-a87e-a299d6bff704.png)

just view and labels so I have full control over the look & feel and the position and very easy/quick to setup

**Reply by beppo-ivel** [reply]
> just view and labels so I have full control over the look & feel and the position and very easy/quick to setup

If I recreate some components for instance to get the "move text up" feature in textfields as we talked about [here](https://github.com/appcelerator/titanium_mobile/discussions/13080). You mentioned I can build that myself with labels, views and a plain textfield. If I do so I would like to implement some reusable code snippets. How do u organize the code  in "templates" to use your own textfield where ever u want? For example, I would start with a 
/views/customComponents/textfield.xml:
```
<Alloy>
  <TextField id="customTxtField" top="10"/>
  <Label id="customLabel" top="5" text="Hello" visible="false"/>
  <Label id="customHintText" top="15" text="Hello"/>
</Alloy>
```
Now I can reuse that snippet via '<Require src="/components/textfield" />' in a arbitrary view and have access to those elements in the corresponding controller, where I can define the functionality e.g.
```
$.customTxtField.addEventListener('focus', function(e) {
  $.customLabel.visible = true;
  $.customHintText.visible = false;
});
```

But to use the snippets more general:
1) I would like to define the functionality once in a "lib/customComponents/textfield.js" file. Then I can import that functionality when ever I need it via require('/components/textfield'); 
2) To make textfield.xml flexible, I have to define  the properties "top" and "text" as variable, because not every textfield has same position and hinttext. 

Can u please show me the idea, how u would organize it, if possible.

**Reply by m1ga** [reply]
Have a look at `Alloy Widgets`: https://titaniumsdk.com/guide/Alloy_Framework/Alloy_Guide/Alloy_Widgets.html
Those are reusablel UI components. Basically you have an own view/styles/controller group which you can use with `<Widget src="foo" name="button"/>`

**Reply by beppo-ivel** [reply]
Thanks, widgets answers my question if I combine it with passing custom arguments. While reading about widgets I a came across custom tags that answers my question as well. I also read that, require views and controllers into your application is an alternative to widget. And I also see the mistake  that I have to put `$.customTxtField.addEventListener('` in a separate controller file and not somewhere under library to provide the accessibility to that view.

**Reply by m1ga** [reply]
Sorry I don't understand it :smile: But I think you've solved it. What I do is export the addEventlistener function in my widget so I can access it from the parent. It is a normal function so you can do `exports.addEventListener = function(evt, clb)[ $.widgetView.addEventListener(evt,clb);});` in your widget.

**Reply by beppo-ivel** [reply]
Let me explain it by using code snippets. When I said "...require views and controllers into your application..." I mean that an alternative to Widgets is to create controllers / views within your project and require them into your application.

views/components/textfield.xml
```
<Alloy>
  <TextField id="customTxtField"/>
  <Label id="customLblMoveUp" visible="false"/>
  <Label id="customLblHintText"/>
</Alloy>
```
controllers/components/textfield.js
```
(function() {
  $.customTxtField.top = $.args.customTop;
  $.customLblHintText.top = $.args.customTop + 17;
  $.customLblMoveUp.top = $.args.customTop - 7;
  $.customLblHintText.text = $.args.customHintText;
  $.customLblMoveUp.text = " " + $.args.customHintText + " ";
})();

$.customTxtField.addEventListener('focus', function(e) {
  $.customLblMoveUp.visible = true;
  $.customLblHintText.visible = false;
});

$.customLblHintText.addEventListener('click', function(e) {
  $.customTxtField.focus();
});

$.customTxtField.addEventListener('blur', function(e) {
  if($.customTxtField.value == ""){
    $.customLblMoveUp.visible = false;
    $.customLblHintText.visible = true;
  }
});
```
I have to combine the require-statement for the view with "passing custom arguments", because I need flexibility at least for position and hint text.` <Require src="/components/textfield" customTop="10" customHintText="MyTextField"/>`. And I have access in the corresponding controller to my custom textfield via Alloy.createController('createTextfield').getView()...; 
There is one main thing left, what to do if you want to add this textfield multiple times in the same view/window? In this case, I think, I have to use custom tags because I don't see nice solution by using widgets or require-statement. 

BTW: can u please explain me how to know the correct relative positions, e.g. that the hint text label is placed in the vertical-middle of the textfield?

**Reply by m1ga** [reply]
I still don't get it, sorry. This is exactly what a Widget is for and there are existing ones like https://github.com/NadsonFernando/input-materialdesign-widget
Perhaps look at that. You can pass data with $.args and if you want multiple you just set an ID. Add methods to them to get/valid stuff and so on. And we are way off-topic in this thread so If you have other questions please open another thread.
if you set height to Ti.UI.SIZE and don't set a top/bottom value the label will be placed in the middle. Otherwise use the postlayout to get the height of the textfield (if it doen't have a fixed height, then you can just use it right away) and set a fixed top or bottom value to the hinttext.
